File: servo_tuning/head-sdk-face/head-server-web/server_utils/simple_yaml.py
from __future__ import annotations


def _parse_scalar(raw_value: str):
    value = raw_value.strip()

    if (
        (value.startswith('"') and value.endswith('"'))
        or (value.startswith("'") and value.endswith("'"))
    ):
        return value[1:-1]

    lowered = value.lower()
    if lowered == "true":
        return True
    if lowered == "false":
        return False
    if lowered == "null":
        return None

    try:
        return int(value)
    except ValueError:
        pass

    try:
        return float(value)
    except ValueError:
        return value


def parse_simple_yaml(text: str = "") -> dict:
    root: dict = {}
    stack: list[tuple[int, dict]] = [(-1, root)]

    for raw_line in text.replace("\r\n", "\n").replace("\r", "\n").split("\n"):
        if not raw_line.strip() or raw_line.strip().startswith("#"):
            continue
        if "\t" in raw_line:
            raise ValueError("Tabs are not supported in YAML config")

        indent = len(raw_line) - len(raw_line.lstrip(" "))
        trimmed = raw_line.strip()
        separator_index = trimmed.find(":")
        if separator_index <= 0:
            raise ValueError(f"Invalid YAML line: {trimmed}")

        key = trimmed[:separator_index].strip()
        raw_value = trimmed[separator_index + 1 :]

        while len(stack) > 1 and indent <= stack[-1][0]:
            stack.pop()

        parent = stack[-1][1]
        if not isinstance(parent, dict):
            raise ValueError(f"Invalid YAML nesting near: {trimmed}")

        if not raw_value.strip():
            child: dict = {}
            parent[key] = child
            stack.append((indent, child))
            continue

        parent[key] = _parse_scalar(raw_value)

    return root

File: servo_tuning/head-sdk-face/head-server-web/viewer/digital/src/simpleYaml.js
function parseScalar(rawValue) {
  const value = rawValue.trim();

  if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith('\'') && value.endsWith('\''))) {
    return value.slice(1, -1);
  }

  if (value === 'true') return true;
  if (value === 'false') return false;
  if (value === 'null') return null;
  if (/^-?\d+(?:\.\d+)?$/.test(value)) return Number(value);

  return value;
}

export function parseSimpleYaml(text = '') {
  const root = {};
  const stack = [{ indent: -1, value: root }];
  const lines = text.replace(/\r\n?/g, '\n').split('\n');

  for (const rawLine of lines) {
    if (!rawLine.trim() || rawLine.trim().startsWith('#')) continue;
    if (rawLine.includes('\t')) {
      throw new Error('Tabs are not supported in remote config YAML');
    }

    const indent = rawLine.match(/^ */)[0].length;
    const trimmed = rawLine.trim();
    const separatorIndex = trimmed.indexOf(':');

    if (separatorIndex <= 0) {
      throw new Error(`Invalid YAML line: ${trimmed}`);
    }

    const key = trimmed.slice(0, separatorIndex).trim();
    const rawValue = trimmed.slice(separatorIndex + 1);

    while (stack.length > 1 && indent <= stack[stack.length - 1].indent) {
      stack.pop();
    }

    const parent = stack[stack.length - 1]?.value;
    if (!parent || typeof parent !== 'object' || Array.isArray(parent)) {
      throw new Error(`Invalid YAML nesting near: ${trimmed}`);
    }

    if (!rawValue.trim()) {
      const child = {};
      parent[key] = child;
      stack.push({ indent, value: child });
      continue;
    }

    parent[key] = parseScalar(rawValue);
  }

  return root;
}
